rclcpp/test/test_node.cpp

```diff
@@ -75,6 +75,39 @@ TEST_F(TestNode, get_name_and_namespace) {
   }
 }
 
+/*
+   Testing node construction and destruction.
+ */
+TEST_F(TestNode, subnode_construction_and_destruction) {
+  {
+    auto node = std::make_shared<rclcpp::Node>("my_node", "/ns");
+    auto subnode = node->create_sub_node("sub_ns");
+  }
+
+  {
+    ASSERT_THROW({
+      auto node = std::make_shared<rclcpp::Node>("my_node", "/ns");
+      auto subnode = node->create_sub_node("invalid_ns?");
+    }, rclcpp::exceptions::InvalidNamespaceError);
+  }
+}
+
+TEST_F(TestNode, subnode_get_name_and_namespace) {
+  {
+    auto node = std::make_shared<rclcpp::Node>("my_node", "/ns");
+    auto subnode = node->create_sub_node("sub_ns");
+    EXPECT_STREQ("my_node", subnode->get_name());
+    EXPECT_STREQ("/ns/sub_ns", subnode->get_namespace());
+  }
+  {
+    auto node = std::make_shared<rclcpp::Node>("my_node", "/ns");
+    auto subnode = node->create_sub_node("/sub_ns");
+    EXPECT_STREQ("my_node", subnode->get_name());
+    EXPECT_STREQ("/ns/sub_ns", subnode->get_namespace());
```

Comment:
Given that the subnode's namespace is fully qualified, I think this result should be `/sub_ns`. This also matches ROS 1 when a fully qualified name is given as a child `NodeHandle` namespace.

```c++
#include "ros/ros.h"
#include "std_msgs/String.h"

int main(int argc, char ** argv)
{
  ros::init(argc, argv, "name");

  ros::NodeHandle n;
  ros::NodeHandle nsub(n, "/sub_ns");
  ros::Publisher pub = nsub4.advertise<std_msgs::String>("nsub/relaive", 1000);

  ros::spin();
  return 0;
}
```

launched with `/path/to/above/executable __ns:=/orig_ns`

Creates the topic `/sub_ns/nsub/relative`, not `/orig_ns/sub_ns/...`

Reply:
We could also consider making it so that "sub nodes" can only use relative name prefix or a prefix starting with a `~`. That would avoid the issue of ignoring the node namespace.